src/assert_safe.h

```diff
@@ -21,28 +21,35 @@
 
 # define GIT_ASSERT(expr) assert(expr)
 # define GIT_ASSERT_ARG(expr) assert(expr)
+
+# define GIT_ASSERT_WITH_RETVAL(expr, fail) assert(expr)
+# define GIT_ASSERT_ARG_WITH_RETVAL(expr, fail) assert(expr)
 #else
 
+/** Internal consistency check to stop the function. */
+# define GIT_ASSERT(expr) GIT_ASSERT_WITH_RETVAL(expr, -1);
+
 /**
  * Assert that a consumer-provided argument is valid, setting an
  * actionable error message and returning -1 if it is not.
  */
-# define GIT_ASSERT_ARG(expr) do { \
-		if (!(expr)) { \
-			git_error_set(GIT_ERROR_INVALID, \
-				"invalid argument: '%s'", \
-				#expr); \
-			return -1; \
-		} \
-	} while(0)
+# define GIT_ASSERT_ARG(expr) GIT_ASSERT_ARG_WITH_RETVAL(expr, -1);
+
+/** Internal consistency check to return the `fail` param on failure. */
+# define GIT_ASSERT_WITH_RETVAL(expr, fail) \
+	GIT_ASSERT__WITH_RETVAL(expr, GIT_ERROR_INTERNAL, "unrecoverable internal error", fail)
+
+/**
+ * Assert that a consumer-provided argument is valid, setting an
+ * actionable error message and returning the `fail` param if not.
+ */
+# define GIT_ASSERT_ARG_WITH_RETVAL(expr, fail) \
+	GIT_ASSERT__WITH_RETVAL(expr, GIT_ERROR_INVALID, "invalid argument", fail)
 
-/* Internal consistency check to stop the function. */
-# define GIT_ASSERT(expr) do { \
+# define GIT_ASSERT__WITH_RETVAL(expr, code, msg, fail) do { \
```

Comment:
Not sure if I'm misreading, but it looks to me like `GIT_ASSERT_WITH_RETVAL` has different parameters in the `GIT_ASSERT_HARD` case compared to the `!GIT_ASSERT_HARD` case.

Edit: misread it. There's two underscores here. Easy to miss that